internal/controller/nicconfigurationtemplate_controller.go

```diff
@@ -40,26 +46,121 @@ type NicConfigurationTemplateReconciler struct {
 //+kubebuilder:rbac:groups=configuration.net.nvidia.com,resources=nicdevices,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=configuration.net.nvidia.com,resources=nicdevices/finalizers,verbs=update
 
-// Reconcile is part of the main kubernetes reconciliation loop which aims to
-// move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the NicConfigurationTemplate object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.17.3/pkg/reconcile
+// Reconcile reconciles the NicConfigurationTemplate object
 func (r *NicConfigurationTemplateReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	template := &v1alpha1.NicConfigurationTemplate{}
+
+	err := r.Get(ctx, req.NamespacedName, template)
+	if err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+
+	nodeList := &v1.NodeList{}
+	err = r.List(ctx, nodeList, client.MatchingLabels(template.Spec.NodeSelector))
+	if err != nil {
+		log.Log.Error(err, "Failed to list nodes")
+		return ctrl.Result{RequeueAfter: 5 * time.Second}, err
+	}
+
+	// Create a map to easily look up nodes by name
+	matchingNodes := make(map[string]struct{})
+	for _, node := range nodeList.Items {
+		matchingNodes[node.Name] = struct{}{}
+	}
+
+	matchingNicTypes := make(map[string]struct{})
+	for _, nicType := range strings.Split(template.Spec.NicSelector.NicType, consts.NicTypeSelectorDelimiter) {
```

Comment:
it was probably not clear in the doc, but you can only have a single type specified  no need to support multiple types.
(else we would have made that into a list)